Improve on gg_lens_pop and others

trivialTZ opened the pull request:
1. For gg_lens.py: a. add input as test_area
b. make position_alignment of lens galaxies be randomly generate in the disk of the test area. a. make criteria for a detectable strong that: source  (xs, ys) are the unlensed source position. (source - lens)

///////////////////////////////////////////////////////////////////////////

2. For gg_lens_pop.py: a. For each lens galaxies tested, create a testarea that within the area we can generate the suitable number of source galaxies and test if they can be strong lensing. b. test all the lens galaxies in generated by yml fire.

///////////////////////////////////////////////////////////////////////////

3. skypy_pipeline.py make sky fraction a variable input outside of yaml file ( if using configuration = yaml.load(file, Loader=yaml.Loader), there will be bug)

trivialTZ committed:
Improve on gg_lens_pop and others

1. For gg_lens.py:
a. add input as test_area
b. make position_alignment of lens galaxies be randomly generate in the disk of the test area.
a. make criteria for a detectable strong that: source  (xs, ys) are the unlensed source position. (source - lens)

///////////////////////////////////////////////////////////////////////////

2. For gg_lens_pop.py:
a. For each lens galaxies tested, create a testarea that within the area we can generate the suitable number of source galaxies and test if they can be strong lensing.
b. test all the lens galaxies in generated by yml fire.

///////////////////////////////////////////////////////////////////////////

3. skypy_pipeline.py
make sky fraction a variable input outside of yaml file ( if using configuration = yaml.load(file, Loader=yaml.Loader), there will be bug)

sibirrer commented on sim_pipeline/gg_lens.py:
```diff
@@ -10,7 +10,7 @@ class GGLens(object):
     """
 
-    def __init__(self, source_dict, deflector_dict, cosmo):
+    def __init__(self, source_dict, deflector_dict, cosmo, test_area=90 * np.pi):
```
can you specify what 'test_area' means here?

trivialTZ replied:
meaning the area of position around lensing galaxies that source galaxies inside the position need to be tested.

trivialTZ replied:
so that the source galaxies' position can be generate within the area.

sibirrer commented on sim_pipeline/Pipelines/skypy_pipeline.py:
```diff
@@ -1,33 +1,42 @@
 import os
 from skypy.pipeline import Pipeline
 import sim_pipeline
+import tempfile
```
'tempfile' is a separate library. can you add it to the requirements.txt file?

trivialTZ replied:
tempfile is a built-in module, no need to install it via pip I think.

sibirrer commented on sim_pipeline/gg_lens_pop.py:
```diff
@@ -72,13 +74,18 @@ def draw_population(self):
-        print(num_lens_systems)
-        for _ in range(100):
-            source = self._source_galaxies.draw_galaxy()
+        num_sources_tested_mean = (testarea * self.f_sky * num_sources)/12960000
```
I don't understand this line. Can you comment what happens here?

trivialTZ replied:
to show the mean of source galaxies needed to tested within the test area.  (later draw a normal distribution around the mean for the number of source galaxies tested);
num_sources_tested_mean/ testarea = num_sources/ f_sky;
12960000= (3600)^2, since testarea is in arcsec^2 and f_sky is in deg^2.

trivialTZ replied:
f_sky have error: num_sources_tested_mean = (testarea * num_sources)/ (12960000 * self.f_sky)

sibirrer commented on sim_pipeline/gg_lens_pop.py:
```diff
@@ -72,13 +74,18 @@ def draw_population(self):
+        print(np.int(num_lenses*num_sources_tested_mean))
+
+# Draw a population of galaxy-galaxy lenses within the area.
+        for _ in range(num_lenses):
```
num_lenses seems not to be defined. I also don't see a test function for it

trivialTZ replied:
![image](https://user-images.githubusercontent.com/122687108/235315381-99a4f05f-2ddb-4e0f-8b7d-f92677bb1baa.png)
        num_lenses = self._lens_galaxies.deflector_number()
I think it is defined, I will add test function about it.

sibirrer commented on sim_pipeline/gg_lens_pop.py:
```diff
@@ -72,13 +74,18 @@ def draw_population(self):
-            # Check the validity of the lens system
-            if gg_lens.validity_test():
-                gg_lens_population.append(gg_lens)
+            for _ in range(max(1,int(np.rint(np.random.normal
```
also here, it is unclear what is the range. Perhaps have a separate function/line of code defining the length of this for-loop and comment on how you calculate it

sibirrer reviewed (changes requested):
Hi @trivialTZ,
I have some comments about documenting/describing of what's happening in the code as I can not fully follow all the steps.
Don't worry about the current CI test failure. It has to do with coveralls, which is not the tests themselves. Seems like their server is down to displace test coverage reports.
Please also add test functions to those routines.

trivialTZ committed:
add test function and other minor updates:

1. requirements add tempfile
2. change default of test_area

trivialTZ committed:
Update requirements.txt

no need to add tempfile since it is a built-in module for python

trivialTZ committed:
add max_image_separation criteria and other minor changes

1.add max_image_separation
2.debug for validity_test 's test of SL images numbers
3.lower the default test area

sibirrer commented on sim_pipeline/gg_lens.py:
```diff
@@ -81,15 +81,24 @@ def validity_test(self, min_image_separation=0, max_image_separation=10):
+            max_image_separation = np.max(separations)
+        if max_image_separation < min_image_separation:
+            return False
+        if max_image_separation > max_image_separation:
```
I am not sure this line makes much sense

trivialTZ replied:
yeah, this is a typo, thank you so much!

sibirrer commented on sim_pipeline/gg_lens.py:
```diff
@@ -81,15 +81,24 @@ def validity_test(self, min_image_separation=0, max_image_separation=10):
+        if len(image_positions[0]) < 2:
             return False
 
+        if len(image_positions[0]) == 2:
```
do you need to distinguish the case of two images vs arbitrary number of images? You might be able to do it with arrays of arbitrary lengths only

trivialTZ replied:
since majority of the images number is 2, I believe if it is did in this way not recall numpy may increase the computer speed. @sibirrer

sibirrer reviewed (changes requested):
Thanks for the update!
I made two comments and can you add a test function for some of the features/individual definitions. Thanks!

trivialTZ committed:
link test_area with Einstein radius at infinity and other minor update

1.add test function for test_area and num_sources_range
2.make methods:image_separation_from_positions; theta_e_when_source_infinity; draw_test_area

sibirrer commented on sim_pipeline/gg_lens.py:
```diff
@@ -5,6 +5,34 @@
+    """
+    calculate image separation in arc-seconds
+
+    :param image_positions: list of image positions in arc-seconds
```
can you be more specific? in the code, you call ten(image_positions[0]) ==2: meaning the input is different than what you stated.

sibirrer commented on sim_pipeline/gg_lens.py:
```diff
@@ -5,6 +5,34 @@
+    return image_separation
+
+
+def theta_e_when_source_infinity(deflector_dict):
```
I would recommend having as input vel_disp directly not relying on a dictionary with a specific keyword argument. This way you can also specify that the input needs to be in km/s.

sibirrer replied:
can you write a specific test function for this definition?

sibirrer commented on sim_pipeline/gg_lens_pop.py:
```diff
@@ -65,7 +75,21 @@ def get_num_lenses(self):
+
+        """
+        num_sources = self._source_galaxies.galaxies_number()
+        num_sources_tested_mean = (testarea * num_sources) / (12960000 * self.f_sky)
```
here you are combining two different things. One is the average source for the render area, and the other is drawing a realization of the expected distribution. I would put the poisson draw in the other function with the strong lensing realization instead of this one

sibirrer commented on sim_pipeline/gg_lens.py:
```diff
@@ -5,6 +5,34 @@
 from lenstronomy.LensModel.Solver.lens_equation_solver import LensEquationSolver
 
 
+def image_separation_from_positions(image_positions):
```
can you write a specific test function for this definition?

sibirrer commented on sim_pipeline/gg_lens_pop.py:
```diff
@@ -95,14 +112,13 @@ def draw_population(self, testarea=4 * np.pi):
+            test_area = draw_test_area(deflector=lens)
+            num_sources_range = self.get_num_sources_tested(testarea=test_area)
+            if num_sources_range > 0:
+                for _ in range(num_sources_range):
```
here we have to be careful. You effectively make a set of single-plane lens systems when instead when num_sources_range > 1 we would expect a multi-source plane lens system (not implemented yet). You can keep it for now but write a TODO that this is not exactly what we would expect

sibirrer reviewed (changes requested):
This looks great. I give some suggestions and minor request for changes and test functions.
Curious to hear whether this change in implementation leads to a speed up or not.

sibirrer commented on sim_pipeline/gg_lens_pop.py:
```diff
@@ -1,5 +1,16 @@
+    :return: test area in arcsec^2
+    """
+    theta_e_infinity = theta_e_when_source_infinity(deflector)
+    test_area = np.pi * (theta_e_infinity*2) ** 2
```
perhaps you can stretch the test area by a tiny bit (say a factor of 1.5 such to make sure that for slightly different mass profiles, also the very edge of two image pairs can form?

sibirrer replied:
ah yes, never mind. I think 1.5 or even 1.2 times the Einstein radius could be enough already. This could lead to some more speedup

trivialTZ commented:
I believe I already stretch it make it 4 times lager. theta_e_infinity*2 am
I correct？

Simon Birrer ***@***.***>于2023年5月4日 周四19:57写道：

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In sim_pipeline/gg_lens_pop.py
> <https://github.com/sibirrer/sim-pipeline/pull/3#discussion_r1185609710>:
>
> > @@ -1,5 +1,16 @@
>  from sim_pipeline.Pipelines.skypy_pipeline import SkyPyPipeline
> -from sim_pipeline.gg_lens import GGLens
> +from sim_pipeline.gg_lens import GGLens, theta_e_when_source_infinity
> +import numpy as np
> +
> +def draw_test_area(deflector):
> +    """
> +    draw a test area around the deflector
> +    :param deflector: deflector dictionary
> +    :return: test area in arcsec^2
> +    """
> +    theta_e_infinity = theta_e_when_source_infinity(deflector)
> +    test_area = np.pi * (theta_e_infinity*2) ** 2
>
> perhaps you can stretch the test area by a tiny bit (say a factor of 1.5
> such to make sure that for slightly different mass profiles, also the very
> edge of two image pairs can form?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sibirrer/sim-pipeline/pull/3#pullrequestreview-1414056861>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5IA5BGHQLVZLHAYPWQTWN3XEQ67LANCNFSM6AAAAAAXP52TBI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

trivialTZ committed:
Updates but error on test function!!!

1.I update code for separating some methods.
2 change the fake data for testing function and tried many other ways but still cannot solved bug about the test function on image_positions.
